Adding and Removing Named Plans. Deprecating Transaction and adding plans directly to the Scheduler.

seanoshea opened the pull request:
Fixes https://github.com/material-motion/material-motion-runtime-android/issues/52

cc @pingpongboss @jverkoey

seanoshea committed:
Per issue #34 - A start on named plans.

seanoshea committed:
Per issue #34 - Should be a NamedPlan when adding it.

seanoshea committed:
Per issue #34 - Adding some more unit tests and some documentation too.

seanoshea committed:
Per issue #34 - Some more tests.

seanoshea committed:
Per issue #34 - Forcing some errors to be thrown and testing for them.

jverkoey commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -85,6 +86,21 @@ void commitPlan(PlanInfo plan) {
 
+  void commitAddNamedPlan(PlanInfo planInfo) {
+    // remove first
+    Performer performer = getNamedPerformer(planInfo);
```
This should be capable of returning null in the event that no performer is presently associated with this name.

seanoshea replied:
Altered `getNamedPerformer` to create the performer if it isn't found.

jverkoey commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -118,6 +134,23 @@ private int getDetailedState() {
+  private Performer getNamedPerformer(PlanInfo planInfo) {
+    Performer namedPerformer = namedCache.get(planInfo.name);
+    if (namedPerformer == null) {
+      namedPerformer = getPerformer(planInfo);
```
Shouldn't be creating a performer if no performer is associated with the name.

pingpongboss replied:
Yes - this should not delegate to `getPerformer()`. You should be able to remove this entire method and just inline `namedCache.get(name)`.

seanoshea replied:
Altered this method to create the performer if it isn't in the cache.

jverkoey commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -85,6 +86,21 @@ void commitPlan(PlanInfo plan) {
+    Performer performer = getNamedPerformer(planInfo);
+    removeNamedPlan(planInfo, performer);
+    // then add
+    if (performer instanceof Performer.NamedPlanPerformance) {
```
Perhaps we want to throw an assertion here if the performer doesn't implement named plan support?

seanoshea replied:
I think this would be taken care of by the fact that a `ClassCastException` would be thrown from https://github.com/seanoshea/material-motion-runtime-android/blob/issues/34-named-plans/library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java#L147 if the performer for this named plan did not conform to `Performer.NamedPlanPerformance` ?

Would you suggest a different type of exception here? I guess we could throw an `IllegalArgumentException` here with a more descriptive error message if you think that'd be useful?

pingpongboss commented on library/src/androidTest/java/com/google/android/material/motion/runtime/SchedulerTest.java:
```diff
@@ -179,6 +341,62 @@ private EndingDelegatedPlan(String text) {
+    @Override
+    public void removePlan(NamedPlan plan, String name) {
+      TextView target = getTarget();
+      target.setText(target.getText() + "removePlanInvoked");
```
pretty clever way to test 👍 

I would add a space in front like `" removePlanInvoked"` so that the assertions are easier to read: 

```
assertTrue(textView.getText().equals(" removePlanInvoked addPlanInvoked"));
```

seanoshea replied:
Sure thing!

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/NamedPlan.java:
```diff
@@ -0,0 +1,33 @@
+  public NamedPlan clone() {
+    return (NamedPlan) super.clone();
+  }
+}
```
I believe main point of this class is to enforce that its Performer is of type NamedPlanPerformance, so you'd want to add:

```
@Override
public abstract Class<? extends NamedPlanPerformer> getPerformerClass();
```

This is to force that contract. (this works in my head. I hope the java generics gods are feeling generous and allow you to actually do this)

seanoshea replied:
Tried:

```

  @Override
  public abstract Class<? extends Performer.NamedPlanPerformance> getPerformerClass();
```

but received this as a compilation error:

```
/Users/sean.oshea/github/material-motion-runtime-android/library/src/main/java/com/google/android/material/motion/runtime/NamedPlan.java:36: error: getPerformerClass() in NamedPlan cannot override getPerformerClass() in Plan
  public abstract Class<? extends Performer.NamedPlanPerformance> getPerformerClass();
                                                                  ^
  return type Class<? extends NamedPlanPerformance> is not compatible with Class<? extends Performer>
/Users/sean.oshea/github/material-motion-runtime-android/library/src/main/java/com/google/android/material/motion/runtime/NamedPlan.java:35: error: method does not override or implement a method from a supertype
  @Override
```

Can look into this some more, but doesn't work immediately out the gate.

pingpongboss replied:
Ohh man this sucks. This would work if `NamedPlanPerformance` is-a `Performer`, but that is not the case since Performer is an abstract class and NamedPlanPerformance is an interface. In the future I can look into whether Performer can become an interface. Don't worry about it in this PR.

pingpongboss replied:
Since this is not possible at the moment, please document that getPerformerClass() should return a Performer that implements NamedPlanPerformance.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/NamedPlan.java:
```diff
@@ -0,0 +1,33 @@
+  @Override
+  public NamedPlan clone() {
+    return (NamedPlan) super.clone();
+  }
```
don't believe this is necessary for any reason - did you override `clone()` on purpose?

seanoshea replied:
Judging by your comment below, I believe we're going to keep this method as it stands today?

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/NamedPlan.java:
```diff
@@ -0,0 +1,33 @@
+package com.google.android.material.motion.runtime;
+
+/**
+ * Similar to {@link Plan} in that a named plan is an object representing what you want something to do.
```
"Similar to" is not completely correct, since NamedPlan is-a Plan. I would also remove much of the explanation of how NamedPlan and Plan act in the same way. That would reduce the javadoc down to:

```
/**
 * A named plan is a {@link Plan} whose performer supports adding and remove the plan by name. Register a named plan by calling {@link Scheduler#addNamedPlan}, and remove it by calling {@link Scheduler#removeNamedPlan}. 
 *
 * A named plan or family of named plans enables fine configuration of a performer's behavior.
 *
 * @see <a>spec</a> 
 */
```

seanoshea replied:
Yep - I like this.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Performer.java:
```diff
@@ -50,6 +50,27 @@
   /**
+   * {@link Performer}s can implement this interface to understand when named plans are added to
+   * or removed from this performer.
+   */
```
for consistency, please closely follow the documentation for `PlanPerformance`.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Performer.java:
```diff
@@ -50,6 +50,27 @@
+  public interface NamedPlanPerformance {
+
+    /**
+     * Executed when a {@link NamedPlan} is added to this Performer.
```
for consistency, please start this method javadoc with a present-tense verb. Also see `PlanPerformance#addPlan` for structure.

"Provides a {@link NamedPlan} to this Performer. The Performer is expected to first remove the effects of any prior plan with the same name, then execute the new plan."

Same for method below.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Scheduler.java:
```diff
@@ -172,6 +172,38 @@ public void addPlan(Plan plan, Object target) {
   }
 
+  /**
+   * Adds a {@link NamedPlan} to this scheduler.
```
it's important to document the side effect that a prior plan with the same name will be first removed.

seanoshea replied:
This should be expanded on in the latest commit. github doesn't seem to be picking up on that though? Happy to make additional changes here though if you think my description of what actually happens need some more work.

seanoshea commented on library/src/main/java/com/google/android/material/motion/runtime/NamedPlan.java:
```diff
@@ -0,0 +1,33 @@
+public abstract class NamedPlan extends Plan implements Cloneable {
+
+  @Override
+  public NamedPlan clone() {
```
Not 100% sure whether this `clone` method is definitely needed here. `Plan` implements `Cloneable` but casts the response from `super.clone()` as a `Plan`. Thought `NamedPlan` would need it's own separate `clone` method, but comments definitely welcome here in case I'm mistaken.

pingpongboss replied:
Yea, the whole Cloneable interface is a bit of black magic. Here's how things worked prior to this PR. `Object.clone()` guarantees that the cloned object is of the same type. So a NamedPlan is cloned to a NamedPlan (even though `Object#clone` returns an `Object`), which is casted by `Plan#clone`'s implementation and returned as a Plan. The only reason Plan overrides `#clone` is for the scheduler to be able to write the following without casting:

```
Plan clone = plan.clone();
```

You would only need to override+cast in NamedPlan if the scheduler wants to write the following:

```
NamedPlan clone = namedPlan.clone();
```

And now that I think about it, you do potentially want this in Scheduler, in which case you should override+cast as you've done.

seanoshea replied:
Yep. Overriding `clone` allows me to do the following in the Scheduler:

```
getTargetScope(target).commitAddNamedPlan(plan.clone(), name, target);
```

I get compilation errors saying that I'm not passing the correct type if I remove this `clone` method from `NamedPlan`

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Scheduler.java:
```diff
@@ -172,6 +172,38 @@ public void addPlan(Plan plan, Object target) {
+   * @param target the target on which the plan will operate.
+   */
+  public void addNamedPlan(NamedPlan plan, String name, Object target) {
+    if (name == null || name.length() == 0) {
```
more idiomatic to write `(name == null || name.isEmpty())`

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Scheduler.java:
```diff
@@ -172,6 +172,38 @@ public void addPlan(Plan plan, Object target) {
+   */
+  public void addNamedPlan(NamedPlan plan, String name, Object target) {
+    if (name == null || name.length() == 0) {
+      throw new AssertionError("A NamedPlan must have a name with more than zero characters");
```
throw IllegalArgumentException

seanoshea replied:
I was looking for something similar to `NSParameterAssert` in Java and settled on `AssertionError` - I think `IllegalArgumentException` gives a better indication to the client about what went wrong though. Thanks!

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Scheduler.java:
```diff
@@ -172,6 +172,38 @@ public void addPlan(Plan plan, Object target) {
+   * @param name the name by which the named plan can be identified.
+   * @param target the target on which the named plan was added.
+   */
+  public void removePlanNamed(String name, Object target) {
```
same changes in this method please

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Scheduler.java:
```diff
@@ -172,6 +172,38 @@ public void addPlan(Plan plan, Object target) {
+    PlanInfo planInfo = new PlanInfo();
+    planInfo.name = name;
+    planInfo.target = target;
+    getTargetScope(target).commitRemoveNamedPlan(planInfo);
```
hm I don't like (mis?)using a partial PlanInfo to just hold the name like that. I would rather see the api be `commitRemoveNamedPlan(name)`

seanoshea replied:
Hmmm. I've made a change to pass along the name, plan and target in some of these API calls which basically allows me to not create a `PlanInfo` object in the `Scheduler`. Not sure if this is the idea you had in mind when you suggested this change, but that's where I've ended up. If we're thinking about moving away from `Transactions`, I think moving away from `PlanInfo` objects mightn't be a bad idea at this stage?

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Scheduler.java:
```diff
@@ -172,6 +172,38 @@ public void addPlan(Plan plan, Object target) {
+   * @param name the name by which the named plan can be identified.
+   * @param target the target on which the named plan was added.
+   */
+  public void removePlanNamed(String name, Object target) {
```
More idiomatic to name this `removeNamedPlan()` in android.

seanoshea replied:
Yep - sure thing!

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -47,6 +47,7 @@
 
   private final SimpleArrayMap<Class<? extends Performer>, Performer> cache =
     new SimpleArrayMap<>();
+  private final SimpleArrayMap<String, Performer> namedCache = new SimpleArrayMap<>();
```
this can be a `SAMap<String, NamedPlanPerformance>` type, which will help you below.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -118,6 +134,23 @@ private int getDetailedState() {
+    }
+  }
+
+  private Performer getNamedPerformer(PlanInfo planInfo) {
```
return value can be NamedPlanPerforming

seanoshea replied:
Yup - I was being a little too generic here.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -118,6 +134,23 @@ private int getDetailedState() {
     return state;
   }
 
+  private void removeNamedPlan(PlanInfo planInfo, Performer performer) {
```
this ties into a previous comment - i would rather not see `PlanInfo` as a param type.

seanoshea committed:
Per issue #34 - Addressing some code review comments. Not creating
PlanInfo objects in the Scheduler anymore.

seanoshea commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -118,24 +133,42 @@ private int getDetailedState() {
+  private Performer.NamedPlanPerformance getNamedPerformer(NamedPlan plan, String name, Object target) {
+    Performer.NamedPlanPerformance namedPerformer = namedCache.get(name);
+    if (namedPerformer == null) {
+      // create it
```
Not 100% sure if this is the correct thing to do here. Should we _always_ create a performer here? Perhaps I should at least check to make sure that the `plan` parameter is not null before proceeding?

seanoshea committed:
Per issue #34 - Adding some Javadoc comments and some more unit tests.

seanoshea committed:
Per issue #34 - Cleanup.

seanoshea commented:
Anything else that I need to take care of here @jverkoey @pingpongboss . I think I've addressed your comments thus far, but if there's anything else that needs to be taken care of, just shout.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -47,6 +47,7 @@
 
   private final SimpleArrayMap<Class<? extends Performer>, Performer> cache =
     new SimpleArrayMap<>();
+  private final SimpleArrayMap<String, Performer.NamedPlanPerformance> namedCache = new SimpleArrayMap<>();
```
throughout PR, prefer `import Performer.NamedPlanPerformance` so you can avoid having the use the fully-qualified name in code. Please change here and in other files.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -118,24 +132,42 @@ private int getDetailedState() {
 
+  private void removeNamedPlan(NamedPlan plan, String name, Performer.NamedPlanPerformance performer) {
+    if (performer != null) {
+      performer.removePlan(plan, name);
```
remove `plan` param from `NamedPlanPerformance#removePlan()` and this `removeNamedPlan()` method, according to spec https://material-motion.gitbooks.io/material-motion-starmap/content/specifications/runtime/named-plans.html

The expectation is that the Performer will keep track of its own named Plans, if necessary (we may change this in the future, but remove for now).

seanoshea replied:
:facepalm: I should have caught this, but went on autopilot and just was using the `addNamedPlan` as a template for this interface.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -118,24 +132,42 @@ private int getDetailedState() {
+    }
+  }
+
+  private Performer.NamedPlanPerformance getNamedPerformer(NamedPlan plan, String name, Object target) {
```
prefer `getPerformerForName()`

EDIT: This method should be going away.

seanoshea replied:
I tried removing this method & just always doing the lookup in the `namedCache`, but the unit tests failed with errors like:

```
com.google.android.material.motion.runtime.SchedulerTest > testRemoveNamedPlanThatWasNeverAdded[Nexus_6_API_23(AVD) - 6.0] FAILED 
    java.lang.NullPointerException: Attempt to invoke interface method 'void com.google.android.material.motion.runtime.Performer$NamedPlanPerformance.addPlan(com.google.android.material.motion.runtime.NamedPlan, java.lang.String)' on a null object reference
    at com.google.android.material.motion.runtime.TargetScope.commitAddNamedPlan(TargetScope.java:96)
:library:connectedDebugAndroidTest FAILED
```

I was thinking `getPerformerForName()` would be useful to wrap this logic:

```
1. Try to look up the performer in namedCache
2. If I cant find it in namedCache, try to create it
3. If I have to create the performer, remember to put it into namedCache
```

Are you suggesting that the `commitAddNamedPlan` method use logic as described above? And, `commitRemoveNamedPlan` would only ever try to look up the performer in `namedCache` ?

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Scheduler.java:
```diff
@@ -172,6 +172,33 @@ public void addPlan(Plan plan, Object target) {
+  /**
+   * Adds a {@link NamedPlan} to this scheduler.
+   * When this method is invoked, a {@link NamedPlan} with the same name and target
+   * is removed from the scheduler before the plan is eventually added.
```
it looks like the latest commit does include your new documentation. Github PR just has a confusing UI when you click on a comment that was made on a previous commit.

seanoshea committed:
Per issue #34 - Code review comments. Removing the named plan from the
removePlan callback.

seanoshea committed:
Per issue #34 - Code review feedback. Will not create a performer for a
`removeName` operation unless the performer has already been created as
a result of `addNamedPlan` being invoked on the scheduler.

seanoshea committed:
Per issue #34 - Code review comments.

seanoshea committed:
Per issue #34 - Merging from upstream.

seanoshea committed:
Merging from develop.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/Transaction.java:
```diff
@@ -60,7 +60,10 @@ public void addPlan(Plan plan, Object target) {
+   * @deprecated  Named Plans should be added directly to the Scheduler instead of using Transactions. <br />
+   *              use {@link com.google.android.material.motion.runtime.Scheduler#addNamedPlan(NamedPlan, String, Object)} on the Scheduler instead
    */
+  @Deprecated
```
nice! all public methods have been @deprecated. Can you @deprecate the class itself as well?

Also, every @deprecated tag should be followed by the version it was deprecated in. In your case, you should insert `3.0.0` since i'm about to push that release (this change will go live in 3.0.0)

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -70,6 +72,22 @@ void commitPlan(PlanInfo plan) {
+    NamedPlanPerformance namedPerformer = namedCache.get(name);
+    if (namedPerformer == null) {
+      namedPerformer = (NamedPlanPerformance)createPerformer(plan, target);
+    }
```
this is fine for now - can you add a `TODO: refactor getPerformer() and use it here` for me

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -70,6 +72,22 @@ void commitPlan(PlanInfo plan) {
+  void commitAddNamedPlan(NamedPlan plan, String name, Object target) {
+    // remove first
+    commitRemoveNamedPlan(name);
+    // then add
```
let's add an empty line before here

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -70,6 +72,22 @@ void commitPlan(PlanInfo plan) {
+    if (namedPerformer == null) {
+      namedPerformer = (NamedPlanPerformance)createPerformer(plan, target);
+    }
+    namedCache.put(name, namedPerformer);
```
move the cache.put() to the end, and add an empty line before it

seanoshea committed:
Per issue #34 - Trimming this out.

pingpongboss commented on library/src/main/java/com/google/android/material/motion/runtime/TargetScope.java:
```diff
@@ -103,24 +121,31 @@ private int getDetailedState() {
     return state;
   }
 
+  private void removeNamedPlan(String name, NamedPlanPerformance performer) {
```
this helper method is pretty useless now. Inline into `commitRemoveNamedPlan()` please

seanoshea committed:
Per issue #34 - mimicking the travis file at
https://github.com/mutexkid/android-studio-robolectric-example/blob/mast
er/.travis.yml

seanoshea committed:
Per issue #52 - Adding these back.

seanoshea committed:
Code review comments. Deprecating Transaction.

pingpongboss reviewed (approved):
this is awesome

seanoshea committed:
Reverting to working copy from develop.
